docs/content/guides/dagster/migrating-to-pythonic-resources-and-config.mdx

```diff
@@ -0,0 +1,313 @@
+from dagster import InitResourceContext, resource
+
+@resource(config_schema=FancyDbResource.to_config_schema())
+def fancy_db_resource(context: InitResourceContext) -> FancyDbResource:
+    return FancyDbResource.from_resource_context(context)
+
+# old-style resource API still works, but the Pythonic resource is the source of truth
+# for schema information and implementation
+defs = Definitions(
+    # ...
+    resources={"fancy_db": fancy_db_resource.configured({"conn_string": "some_value"})},
+)
+```
+
+Note that we have written a Pythonic resource while still passing the old-style one into <PyObject object="Definitions"/>, and there is a single source of truth for the config schema.
+
+### Step 2: Providing the Pythonic resource to `Definitions`
+
+Now we can change `Definitions` to take the Pythonic resource. We do not need to migrate our asset code in lockstep. Pythonic resources are still put on the context.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_new_style_resource_on_context endbefore=end_new_style_resource_on_context dedent=4
+from dagster import ConfigurableResource, Definitions, OpExecutionContext, asset
+
+class FancyDbResource(ConfigurableResource):
+    conn_string: str
+
+    def execute(self, query: str) -> None:
+        ...
+
+@asset(required_resource_keys={"fancy_db"})
+def asset_one(context: OpExecutionContext) -> None:
+    # this still works because the resource is still available on the context
+    assert context.resources.fancy_db
+
+defs = Definitions(
+    assets=[asset_one],
+    resources={"fancy_db": FancyDbResource(conn_string="some_value")},
+)
+```
+
+### Step 3: Using the resource in assets
+
+Now we can move the asset to take the resource as a parameter.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_new_style_resource_on_param endbefore=end_new_style_resource_on_param dedent=4
+from dagster import OpExecutionContext, asset
+
+@asset
+def asset_one(context: OpExecutionContext, fancy_db: FancyDbResource) -> None:
+    assert fancy_db
+```
+
+## Migrating a resource that uses a separate object for business logic
+
+Frequently users want to separate out the interaction with the configuration system from the object that contains the actual business logic. Maybe it is pre-existing or third-party code that knows nothing about the configuration system and shouldn't. Or it is a complex client that requires mutable state and bookkeeping. In the old API the engineer would have written a `@resource` decorator that returns that object directly.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_old_third_party_resource endbefore=end_old_third_party_resource dedent=4
+# Pre-existing code that you don't want to alter
+class FancyDbClient:
```

Comment:
personal preference, and i know the beginning of the section says "this is the code we'll be referencing the whole time" but i feel like this third party code snippet might benefit from the class names being different than the previous examples. Like `FanycDbClient` -> `PreExistingDbClient` and `from fancy_db import FancyDbClient` -> `from external_fancy_db import ExternalFancyDbClient` or something like that

Reply:
Started down this road but it didn't really sit with me right. Totally get what you are saying.